### PR TITLE
chan_pjsip: Add PJSIPHangup dialplan app and manager action

### PR DESCRIPTION
**Pull request by gtjoseph**
See UserNote below.

Exposed the existing Hangup AMI action in manager.c so we can use
all of it's channel search and AMI protocol handling without
duplicating that code in dialplan_functions.c.

Added a lookup function to res_pjsip.c that takes in the
string represenation of the pjsip_status_code enum and returns
the actual status code.  I.E.  ast_sip_str2rc("DECLINE") returns
603.  This allows the caller to specify PJSIPHangup(decline) in
the dialplan, just like Hangup(call_rejected).

Also extracted the XML documentation to its own file since it was
almost as large as the code itself.

UserNote: A new dialplan app PJSIPHangup and AMI action allows you
to hang up an unanswered incoming PJSIP call with a specific SIP
response code in the 400 -> 699 range.

### CODE REVIEW

**Comment by gtjoseph** [comment]
cherry-pick-to: 18
cherry-pick-to: 20
cherry-pick-to: 21
cherry-pick-to: certified/18.9

**Comment by gtjoseph** [comment]
oops...forgot there needs to be a companion to this that sets a Reason header on the next response.  will do that tomorrow.

File: channels/pjsip/dialplan_functions.c [hunk 1]
```diff
@@ -204,6 +204,25 @@
...
+			<para>SIP response code between 400 and 699.</para>
+		</parameter>
+	</syntax>
+	<description>
+		<para>Hangs up an incoming PJSIP channel and returns the
+		specified SIP response code to the caller.</para>
+		<warning><para>
+		This function must be called BEFORE anything that
+		might cause any other final (non 1XX) response to be sent.
+		</para></warning>
```
Comment (InterLinked1) [inline channels/pjsip/dialplan_functions.c:222]: An example would be nice.

e.g. `exten => s,1,Set(PJSIP_HANGUP()=404)`

And I think people might want to know if they can do other stuff after calling this, will dialplan execution end immediately after?

Are there examples of things that might cause a final response to be sent?

File: channels/pjsip/dialplan_functions.c [hunk 2]
```diff
@@ -1784,3 +1803,65 @@ int pjsip_acf_session_refresh_write(struct ast_channel *chan, const char *cmd, c
...
+{
+	struct ast_sip_channel_pvt *channel;
+	struct hangup_data hdata;
+
+	if (!chan) {
+		ast_log(LOG_WARNING, "No channel was provided to %s function.\n", cmd);
+		return -1;
+	}
+
+	if (sscanf(value, "%30d", &hdata.response_code) != 1) {
```
Comment (InterLinked1) [inline channels/pjsip/dialplan_functions.c:1835]: Isn't ast_str_to_int is preferred over sscanf, for int parsing?

**Review by InterLinked1 — APPROVED** [review]

File: channels/pjsip/dialplan_functions.c [hunk 3]
```diff
@@ -1784,3 +1228,71 @@ int pjsip_acf_session_refresh_write(struct ast_channel *chan, const char *cmd, c
...
+	int response_code;
+};
+
+static int pjsip_hangup(void *obj)
+{
+	struct hangup_data *hdata = obj;
+	pjsip_tx_data *packet = NULL;
+
+	if ((hdata->session->inv_session->state != PJSIP_INV_STATE_DISCONNECTED) &&
+		(pjsip_inv_answer(hdata->session->inv_session, hdata->response_code, NULL, NULL, &packet) == PJ_SUCCESS)) {
```
Comment (jcolp) [inline channels/pjsip/dialplan_functions.c:1243]: What's the difference between using pjsip_inv_answer here and pjsip_inv_end_session?

File: channels/pjsip/dialplan_functions.c [hunk 4]
```diff
@@ -1784,3 +1228,71 @@ int pjsip_acf_session_refresh_write(struct ast_channel *chan, const char *cmd, c
...
+struct hangup_data {
+	struct ast_sip_session *session;
+	int response_code;
+};
+
+static int pjsip_hangup(void *obj)
+{
+	struct hangup_data *hdata = obj;
+	pjsip_tx_data *packet = NULL;
+
```
Comment (jcolp) [inline channels/pjsip/dialplan_functions.c:1241]: The hangup cause on the channel should be a conversion from SIP response code to ISDN cause code for a bit of consistency

File: channels/pjsip/dialplan_functions.xml [hunk 5]
```diff
@@ -0,0 +1,591 @@
...
+			Hangup an incoming PJSIP channel with a SIP response code
+		</synopsis>
+		<syntax>
+			<parameter name="response_code" required="true">
+				<para>SIP response code between 400 and 699.</para>
+			</parameter>
+		</syntax>
+		<description>
+			<para>
+			Hangs up an incoming PJSIP channel and returns the
```
Comment (jcolp) [inline channels/pjsip/dialplan_functions.xml:189]: It should be documented what the behavior of the ISDN cause code is with this

**Review by jcolp — DISMISSED** [review]

File: channels/pjsip/dialplan_functions.c (hunk 3 above)
Comment (gtjoseph) [reply channels/pjsip/dialplan_functions.c:1243]: pjsip_inv_end_session just calls pjsip_inv_answer when we're a UAS and the state is PJSIP_INV_STATE_INCOMING (which we've already tested for).

File: channels/pjsip/dialplan_functions.c (hunk 4 above)
Comment (gtjoseph) [reply channels/pjsip/dialplan_functions.c:1241]: It's set by chan_pjsip_session_end when the session ends.

File: channels/pjsip/dialplan_functions.xml (hunk 5 above)
Comment (gtjoseph) [reply channels/pjsip/dialplan_functions.xml]: Done

**Review by jcolp — APPROVED** [review]

**Comment by jcolp** [comment]
This should also have test coverage in the testsuite.

**Comment by gtjoseph** [comment]
> This should also have test coverage in the testsuite.

Yep.  In progress.

**Pull request merged**